File: qt/gen_qrhiwidget.cpp
// +build ignore

#include <QEvent>
#include <QImage>
#include <QMetaObject>
#include <QObject>
#include <QPaintDevice>
#include <QPaintEvent>
#include <QResizeEvent>
#include <QRhiWidget>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QWidget>
#include <qrhiwidget.h>
#include "gen_qrhiwidget.h"

class MiqtVirtualQRhiWidget : public virtual QRhiWidget {
public:

	MiqtVirtualQRhiWidget(QWidget* parent): QRhiWidget(parent) {};
	MiqtVirtualQRhiWidget(): QRhiWidget() {};
	MiqtVirtualQRhiWidget(QWidget* parent, Qt::WindowFlags f): QRhiWidget(parent, f) {};

	virtual ~MiqtVirtualQRhiWidget() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QRhiWidget::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QRhiWidget_MetaObject(const_cast<MiqtVirtualQRhiWidget*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QRhiWidget::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QRhiWidget::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QRhiWidget_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QRhiWidget::qt_metacast(param1);

	}

};

QRhiWidget* QRhiWidget_new(QWidget* parent) {
	return new MiqtVirtualQRhiWidget(parent);
}

QRhiWidget* QRhiWidget_new2() {
	return new MiqtVirtualQRhiWidget();
}

QRhiWidget* QRhiWidget_new3(QWidget* parent, int f) {
	return new MiqtVirtualQRhiWidget(parent, static_cast<Qt::WindowFlags>(f));
}

void QRhiWidget_virtbase(QRhiWidget* src, QWidget** outptr_QWidget) {
	*outptr_QWidget = static_cast<QWidget*>(src);
}

QMetaObject* QRhiWidget_MetaObject(const QRhiWidget* self) {
	return (QMetaObject*) self->metaObject();
}

void* QRhiWidget_Metacast(QRhiWidget* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QRhiWidget_Tr(const char* s) {
	QString _ret = QRhiWidget::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

Api QRhiWidget_Api(const QRhiWidget* self) {
	return self->api();
}

void QRhiWidget_SetApi(QRhiWidget* self, Api api) {
	self->setApi(api);
}

bool QRhiWidget_IsDebugLayerEnabled(const QRhiWidget* self) {
	return self->isDebugLayerEnabled();
}

void QRhiWidget_SetDebugLayerEnabled(QRhiWidget* self, bool enable) {
	self->setDebugLayerEnabled(enable);
}

int QRhiWidget_SampleCount(const QRhiWidget* self) {
	return self->sampleCount();
}

void QRhiWidget_SetSampleCount(QRhiWidget* self, int samples) {
	self->setSampleCount(static_cast<int>(samples));
}

TextureFormat QRhiWidget_ColorBufferFormat(const QRhiWidget* self) {
	return self->colorBufferFormat();
}

void QRhiWidget_SetColorBufferFormat(QRhiWidget* self, TextureFormat format) {
	self->setColorBufferFormat(format);
}

QSize* QRhiWidget_FixedColorBufferSize(const QRhiWidget* self) {
	return new QSize(self->fixedColorBufferSize());
}

void QRhiWidget_SetFixedColorBufferSize(QRhiWidget* self, QSize* pixelSize) {
	self->setFixedColorBufferSize(*pixelSize);
}

void QRhiWidget_SetFixedColorBufferSize2(QRhiWidget* self, int w, int h) {
	self->setFixedColorBufferSize(static_cast<int>(w), static_cast<int>(h));
}

bool QRhiWidget_IsMirrorVerticallyEnabled(const QRhiWidget* self) {
	return self->isMirrorVerticallyEnabled();
}

void QRhiWidget_SetMirrorVertically(QRhiWidget* self, bool enabled) {
	self->setMirrorVertically(enabled);
}

QImage* QRhiWidget_GrabFramebuffer(const QRhiWidget* self) {
	return new QImage(self->grabFramebuffer());
}

void QRhiWidget_FrameSubmitted(QRhiWidget* self) {
	self->frameSubmitted();
}

void QRhiWidget_connect_FrameSubmitted(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)()>(&QRhiWidget::frameSubmitted), self, [=]() {
		miqt_exec_callback_QRhiWidget_FrameSubmitted(slot);
	});
}

void QRhiWidget_RenderFailed(QRhiWidget* self) {
	self->renderFailed();
}

void QRhiWidget_connect_RenderFailed(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)()>(&QRhiWidget::renderFailed), self, [=]() {
		miqt_exec_callback_QRhiWidget_RenderFailed(slot);
	});
}

void QRhiWidget_SampleCountChanged(QRhiWidget* self, int samples) {
	self->sampleCountChanged(static_cast<int>(samples));
}

void QRhiWidget_connect_SampleCountChanged(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)(int)>(&QRhiWidget::sampleCountChanged), self, [=](int samples) {
		int sigval1 = samples;
		miqt_exec_callback_QRhiWidget_SampleCountChanged(slot, sigval1);
	});
}

void QRhiWidget_ColorBufferFormatChanged(QRhiWidget* self, TextureFormat format) {
	self->colorBufferFormatChanged(format);
}

void QRhiWidget_connect_ColorBufferFormatChanged(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)(TextureFormat)>(&QRhiWidget::colorBufferFormatChanged), self, [=](TextureFormat format) {
		TextureFormat sigval1 = format;
		miqt_exec_callback_QRhiWidget_ColorBufferFormatChanged(slot, sigval1);
	});
}

void QRhiWidget_FixedColorBufferSizeChanged(QRhiWidget* self, QSize* pixelSize) {
	self->fixedColorBufferSizeChanged(*pixelSize);
}

void QRhiWidget_connect_FixedColorBufferSizeChanged(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)(const QSize&)>(&QRhiWidget::fixedColorBufferSizeChanged), self, [=](const QSize& pixelSize) {
		const QSize& pixelSize_ret = pixelSize;
		// Cast returned reference into pointer
		QSize* sigval1 = const_cast<QSize*>(&pixelSize_ret);
		miqt_exec_callback_QRhiWidget_FixedColorBufferSizeChanged(slot, sigval1);
	});
}

void QRhiWidget_MirrorVerticallyChanged(QRhiWidget* self, bool enabled) {
	self->mirrorVerticallyChanged(enabled);
}

void QRhiWidget_connect_MirrorVerticallyChanged(QRhiWidget* self, intptr_t slot) {
	MiqtVirtualQRhiWidget::connect(self, static_cast<void (QRhiWidget::*)(bool)>(&QRhiWidget::mirrorVerticallyChanged), self, [=](bool enabled) {
		bool sigval1 = enabled;
		miqt_exec_callback_QRhiWidget_MirrorVerticallyChanged(slot, sigval1);
	});
}

struct miqt_string QRhiWidget_Tr2(const char* s, const char* c) {
	QString _ret = QRhiWidget::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QRhiWidget_Tr3(const char* s, const char* c, int n) {
	QString _ret = QRhiWidget::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QRhiWidget_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQRhiWidget*>( (QRhiWidget*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QRhiWidget_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQRhiWidget*)(self) )->virtualbase_MetaObject();
}

void QRhiWidget_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQRhiWidget*>( (QRhiWidget*)(self) )->handle__Metacast = slot;
}

void* QRhiWidget_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQRhiWidget*)(self) )->virtualbase_Metacast(param1);
}

void QRhiWidget_Delete(QRhiWidget* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQRhiWidget*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qrhiwidget.go
package qt

import (
	"unsafe"
)

type QRhiWidget__Api int

const (
	QRhiWidget__Null       QRhiWidget__Api = 0
	QRhiWidget__OpenGL     QRhiWidget__Api = 1
	QRhiWidget__Metal      QRhiWidget__Api = 2
	QRhiWidget__Vulkan     QRhiWidget__Api = 3
	QRhiWidget__Direct3D11 QRhiWidget__Api = 4
	QRhiWidget__Direct3D12 QRhiWidget__Api = 5
)

type QRhiWidget__TextureFormat int

const (
	QRhiWidget__RGBA8   QRhiWidget__TextureFormat = 0
	QRhiWidget__RGBA16F QRhiWidget__TextureFormat = 1
	QRhiWidget__RGBA32F QRhiWidget__TextureFormat = 2
	QRhiWidget__RGB10A2 QRhiWidget__TextureFormat = 3
)

type QRhiWidget struct {
	h          uintptr
	isSubclass bool
}

// NewQRhiWidget constructs a new QRhiWidget object.
func NewQRhiWidget(parent *QWidget) *QRhiWidget {
	ret := newQRhiWidget(QRhiWidget_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQRhiWidget2 constructs a new QRhiWidget object.
func NewQRhiWidget2() *QRhiWidget {
	ret := newQRhiWidget(QRhiWidget_new2())
	ret.isSubclass = true
	return ret
}

// NewQRhiWidget3 constructs a new QRhiWidget object.
func NewQRhiWidget3(parent *QWidget, f WindowType) *QRhiWidget {
	ret := newQRhiWidget(QRhiWidget_new3(parent.cPointer(), (int)(f)))
	ret.isSubclass = true
	return ret
}

func (this *QRhiWidget) MetaObject() *QMetaObject {
	return newQMetaObject(QRhiWidget_MetaObject(this.h))
}

func (this *QRhiWidget) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QRhiWidget_Metacast(this.h, param1_Cstring))
}

func QRhiWidget_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QRhiWidget_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QRhiWidget) Api() Api {
	xxxxxxxxx
}

func (this *QRhiWidget) SetApi(api Api) {
	QRhiWidget_SetApi(this.h, api)
}

func (this *QRhiWidget) IsDebugLayerEnabled() bool {
	return (bool)(QRhiWidget_IsDebugLayerEnabled(this.h))
}

func (this *QRhiWidget) SetDebugLayerEnabled(enable bool) {
	QRhiWidget_SetDebugLayerEnabled(this.h, (bool)(enable))
}

func (this *QRhiWidget) SampleCount() int {
	return (int)(QRhiWidget_SampleCount(this.h))
}

func (this *QRhiWidget) SetSampleCount(samples int) {
	QRhiWidget_SetSampleCount(this.h, (int)(samples))
}

func (this *QRhiWidget) ColorBufferFormat() TextureFormat {
	xxxxxxxxx
}

func (this *QRhiWidget) SetColorBufferFormat(format TextureFormat) {
	QRhiWidget_SetColorBufferFormat(this.h, format)
}

func (this *QRhiWidget) FixedColorBufferSize() *QSize {
	_goptr := newQSize(QRhiWidget_FixedColorBufferSize(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QRhiWidget) SetFixedColorBufferSize(pixelSize QSize) {
	QRhiWidget_SetFixedColorBufferSize(this.h, pixelSize.cPointer())
}

func (this *QRhiWidget) SetFixedColorBufferSize2(w int, h int) {
	QRhiWidget_SetFixedColorBufferSize2(this.h, (int)(w), (int)(h))
}

func (this *QRhiWidget) IsMirrorVerticallyEnabled() bool {
	return (bool)(QRhiWidget_IsMirrorVerticallyEnabled(this.h))
}

func (this *QRhiWidget) SetMirrorVertically(enabled bool) {
	QRhiWidget_SetMirrorVertically(this.h, (bool)(enabled))
}

func (this *QRhiWidget) GrabFramebuffer() *QImage {
	_goptr := newQImage(QRhiWidget_GrabFramebuffer(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QRhiWidget) FrameSubmitted() {
	QRhiWidget_FrameSubmitted(this.h)
}

func (this *QRhiWidget) OnFrameSubmitted(slot func()) {
	QRhiWidget_connect_FrameSubmitted(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_FrameSubmitted
func miqt_exec_callback_QRhiWidget_FrameSubmitted(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QRhiWidget) RenderFailed() {
	QRhiWidget_RenderFailed(this.h)
}

func (this *QRhiWidget) OnRenderFailed(slot func()) {
	QRhiWidget_connect_RenderFailed(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_RenderFailed
func miqt_exec_callback_QRhiWidget_RenderFailed(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QRhiWidget) SampleCountChanged(samples int) {
	QRhiWidget_SampleCountChanged(this.h, (int)(samples))
}

func (this *QRhiWidget) OnSampleCountChanged(slot func(samples int)) {
	QRhiWidget_connect_SampleCountChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_SampleCountChanged
func miqt_exec_callback_QRhiWidget_SampleCountChanged(cb intptr_t, samples int) {
	gofunc, ok := cgo.Handle(cb).Value().(func(samples int))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (int)(samples)

	gofunc(slotval1)
}

func (this *QRhiWidget) ColorBufferFormatChanged(format TextureFormat) {
	QRhiWidget_ColorBufferFormatChanged(this.h, format)
}

func (this *QRhiWidget) OnColorBufferFormatChanged(slot func(format TextureFormat)) {
	QRhiWidget_connect_ColorBufferFormatChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_ColorBufferFormatChanged
func miqt_exec_callback_QRhiWidget_ColorBufferFormatChanged(cb intptr_t, format TextureFormat) {
	gofunc, ok := cgo.Handle(cb).Value().(func(format TextureFormat))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	xxxxxxxxx

	gofunc(slotval1)
}

func (this *QRhiWidget) FixedColorBufferSizeChanged(pixelSize *QSize) {
	QRhiWidget_FixedColorBufferSizeChanged(this.h, pixelSize.cPointer())
}

func (this *QRhiWidget) OnFixedColorBufferSizeChanged(slot func(pixelSize *QSize)) {
	QRhiWidget_connect_FixedColorBufferSizeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_FixedColorBufferSizeChanged
func miqt_exec_callback_QRhiWidget_FixedColorBufferSizeChanged(cb intptr_t, pixelSize *QSize) {
	gofunc, ok := cgo.Handle(cb).Value().(func(pixelSize *QSize))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := newQSize(pixelSize)

	gofunc(slotval1)
}

func (this *QRhiWidget) MirrorVerticallyChanged(enabled bool) {
	QRhiWidget_MirrorVerticallyChanged(this.h, (bool)(enabled))
}

func (this *QRhiWidget) OnMirrorVerticallyChanged(slot func(enabled bool)) {
	QRhiWidget_connect_MirrorVerticallyChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_MirrorVerticallyChanged
func miqt_exec_callback_QRhiWidget_MirrorVerticallyChanged(cb intptr_t, enabled bool) {
	gofunc, ok := cgo.Handle(cb).Value().(func(enabled bool))
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	slotval1 := (bool)(enabled)

	gofunc(slotval1)
}

func QRhiWidget_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QRhiWidget_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QRhiWidget_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QRhiWidget_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QRhiWidget) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QRhiWidget_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QRhiWidget) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QRhiWidget_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_MetaObject
func miqt_exec_callback_QRhiWidget_MetaObject(self QRhiWidget, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QRhiWidget{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QRhiWidget) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QRhiWidget_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QRhiWidget) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QRhiWidget_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QRhiWidget_Metacast
func miqt_exec_callback_QRhiWidget_Metacast(self QRhiWidget, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QRhiWidget{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
